osu.Game.Rulesets.Mania/Mods/ManiaModHoldOff.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Mania.Objects;
+using osu.Game.Rulesets.Mods;
+using osu.Framework.Graphics.Sprites;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Mania.Beatmaps;
+
+namespace osu.Game.Rulesets.Mania.Mods
+{
+    public class ManiaModHoldOff : Mod, IApplicableAfterBeatmapConversion
+    {
+        public override string Name => "Hold Off";
+
+        public override string Acronym => "HO";
+
+        public override double ScoreMultiplier => 1;
+
+        public override string Description => @"Turns all hold notes into normal notes. No coordination required.";
+
+        public override IconUsage? Icon => FontAwesome.Solid.DotCircle;
+
+        public override ModType Type => ModType.Conversion;
+
+        public override Type[] IncompatibleMods => new[] { typeof(ManiaModInvert) };
+
+        public const double END_NOTE_ALLOW_THRESHOLD = 0.5;
+
+        public void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            var maniaBeatmap = (ManiaBeatmap)beatmap;
+
+            var newObjects = new List<ManiaHitObject>();
+
+            foreach (var h in beatmap.HitObjects.OfType<HoldNote>())
+            {
+                // Add a note for the beginning of the hold note
+                newObjects.Add(new Note
+                {
+                    Column = h.Column,
+                    StartTime = h.StartTime,
+                    Samples = h.GetNodeSamples(0)
+                });
+
+                // Don't add an end note if the duration is shorter than the threshold
+                double noteValue = GetNoteDurationInBeatLength(h, maniaBeatmap); // 1/1, 1/2, 1/4, etc.
+
+                if (noteValue >= END_NOTE_ALLOW_THRESHOLD)
+                {
+                    newObjects.Add(new Note
+                    {
+                        Column = h.Column,
+                        StartTime = h.EndTime,
+                        Samples = h.GetNodeSamples((h.NodeSamples?.Count - 1) ?? 1)
+                    });
+                }
+            }
+
+            maniaBeatmap.HitObjects = maniaBeatmap.HitObjects.OfType<Note>().Concat(newObjects).OrderBy(h => h.StartTime).ToList();
+        }
+
+        public static double GetNoteDurationInBeatLength(HoldNote holdNote, ManiaBeatmap beatmap)
+        {
+            double bpmAtNoteTime = beatmap.ControlPointInfo.TimingPointAt(holdNote.StartTime).BPM;
+            return (60 * holdNote.Duration) / (1000 * bpmAtNoteTime);
```

Comment:
while this appears to somehow work, it is really complicated for some reason, to the point where i have no idea what the constants are doing - surely it'd be easier to write

```suggestion
            double beatLength = beatmap.ControlPointInfo.TimingPointAt(holdNote.StartTime).BeatLength;
            return holdNote.Duration / beatLength;
```

or is there some subtlety i'm missing? my version appears to pass tests, at least.

Reply:
Looks like a big logical improvement, lol. The original logic was based on the thought that at 60BPM, 1 beat = 1 second, at 120 BPM 1 beat = 0.5 s, etc., so we had (duration*60) / (BPM*1000) = value in beats. As you said, even formulating an explanation for this is rather difficult; I simply arranged the terms until the formula worked, to be frank. It didn't occur to me to use the beatlength, but indeed this ought to be the same. Thanks for the improvement.